web/src/test/java/org/springframework/security/web/firewall/StrictHttpFirewallTests.java

```diff
@@ -595,4 +597,38 @@ public void getFirewalledRequestWhenUntrustedDomainThenException() {
 
 		this.firewall.getFirewalledRequest(this.request);
 	}
+
+	@Test(expected = RequestRejectedException.class)
+	public void getFirewalledRequestWhenControlCharacterInHeaderNameThenException() {
+		this.request.addHeader("Bad\0Name", "some value");
+
+		HttpServletRequest request = this.firewall.getFirewalledRequest(this.request);
+		request.getHeaderNames();
+	}
+
+	@Test(expected = RequestRejectedException.class)
+	public void getFirewalledRequestWhenControlCharacterInHeaderValueThenException() {
+		this.request.addHeader("Something", "bad\0value");
+
+		HttpServletRequest request = this.firewall.getFirewalledRequest(this.request);
+		request.getHeader("Something");
+	}
+
+	@Test(expected = RequestRejectedException.class)
+	public void getFirewalledRequestWhenControlCharacterInParameterNameThenException() {
+		this.request.addParameter("Bad\0Name", "some value");
+
+		HttpServletRequest request = this.firewall.getFirewalledRequest(this.request);
+		request.getParameterNames();
+	}
+
+	@Test(expected = RequestRejectedException.class)
+	public void getFirewalledRequestWhenNotAllowedInParameterValueThenException() {
```

Comment:
Would you please add more tests to confirm that the different setters work? It would also be nice to confirm that the different request getters, e.g. `getParameter`, correctly reject a bad parameter name.

Reply:
Absolutely! I've added a number of tests that I believe address this request.